fecgov/fec-cms
# #2840 Implement Tag Manager

## Summary

- Resolves #2840 

Progress:

- [x] Wrap current analytics inside the feature flag
- [x] Create Tag Manager (GTM) container for non-Production
- [x] Add Production GTM container to pages behind feature flag
- [x] Add non-Production container to pages as an alternate to the Production container
- [x] Add `<noscript>` GTM elements to templates
- [x] Add code to non-Production container to confirm it's loading
- [ ] **Confirm proper GA/DAP/GTM elements are loading**
- [ ] Comb site to migrate any necessary GA events to GTM data layer
- [ ] Confirm successful migration of GA events to GTM
- [ ] Confirm GSA DAP is unaffected (all GSA DAP events are still coming through)
- [ ] Remove current/then-retired GA and DAP implementation
- [ ] Remove feature flag?
- [ ] Add new events?

## Impacted areas of the application

Every page of the site should be affected.

I was getting some linting messages so I let it add some line breaks. Created the env var.

The major changes were to the templates:
- fec/data/templates/layouts/[main, widgets].jinja
- fec/fec/templates/[base, home_base].html

## Screenshots

None

## Related PRs

None

## How to test
- Pull and build the site like usual
- Check the homepage for the GSA DAP tag
  1. Update DEBUG settings to false
  1. Run server
  1. check the page source (in the browser)
  1. cmd-f for `_fed_an_ua_tag` (the GSA DAP element)
  1. confirm that the tag is in the page
  1. cmd-f for 'googletagmanager`
  1. confirm that it does **not** exist
- Visit a data page and confirm that the tag is in the page source
- Go to the calendar to confirm that the GSA DAP element (`_fed_an_ua_tag`) is in the page.
- Interrupt the server
- Set the feature flag `export set FEC_FEATURE_USE_TAG_MANAGER=true`
- Restart the server
- Check the homepage source again. Confirm that:
  1. `_fed_an_ua_tag` is gone
  1. `googletagmanager` **does** appear
  1. GTM is using the `GTM-PMSLNL3` container (search the page source, or check near the end of the GTM `<script>`)
  1. GTM is **not** using the production container of `GTM-T5HPRLH`
  1. the Console in inspector is showing a message like "loaded the Tag Manager dev container"
- Check those same things for a data page
- Check those same things for a calendar page
- Celebrate 🎉

## Next steps

See the Progress list in the Summary at the top
____
